pkg/controller/agentregistration/server.go

```diff
@@ -0,0 +1,150 @@
+package agentregistration
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	authorizationv1 "k8s.io/api/authorization/v1"
+
+	"github.com/stolostron/managedcluster-import-controller/pkg/controller/importconfig"
+	"github.com/stolostron/managedcluster-import-controller/pkg/helpers"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	genericoptions "k8s.io/apiserver/pkg/server/options"
+	"k8s.io/klog/v2"
+)
+
+func NewAgentRegistrationServer(port int, clientHolder *helpers.ClientHolder) *AgentRegistrationServer {
+	return &AgentRegistrationServer{
+		port:         port,
+		clientHolder: clientHolder,
+	}
+}
+
+// AgentRegistrationServer returns a bundle of manifests to the agent to apply for the registration process.
+type AgentRegistrationServer struct {
+	clientHolder *helpers.ClientHolder
+	port         int
+	auth         authenticator.Request
+}
+
+func (s *AgentRegistrationServer) Run(ctx context.Context) error {
+	var err error
+
+	if s.port == 0 {
+		return fmt.Errorf("port is not set")
+	}
+
+	// client auth
+	authConfig := &DelegatingAuthenticatorConfig{
+		CacheTTL:                 10 * time.Second,
+		WebhookRetryBackoff:      genericoptions.DefaultAuthWebhookRetryBackoff(),
+		TokenAccessReviewTimeout: 10 * time.Second,
+	}
+	s.auth, err = authConfig.New()
+	if err != nil {
+		return err
+	}
+	authConfig.Start(ctx)
+
+	server := &http.Server{
+		ReadHeaderTimeout: 10 * time.Second,
+		Addr:              fmt.Sprintf(":%d", s.port),
+		Handler:           s,
+		TLSConfig:         &tls.Config{MinVersion: tls.VersionTLS12},
+	}
+
+	klog.Infof("Starting AgentRegistrationServer on port %d", s.port)
+	return server.ListenAndServeTLS("/server/tls.crt", "/server/tls.key")
+}
+
+func (s *AgentRegistrationServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var err error
+	var content []byte
+
+	klog.Infof("AgentRegistrationServer request: %s %s", r.Method, r.RequestURI)
+
+	// Authenticate the request
+	authResponse, pass, err := s.auth.AuthenticateRequest(r)
```

Comment:
I think you can make it easier by just create a tokereview request to the apiserver. delegation is used only commonly with generic apiserver, it is too heavy here.

Reply:
Agree, replaced it with just a tokenreview request.